// ng-masonry-layout/esm5/lib/ngx-masonry.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Inject, ElementRef, forwardRef, PLATFORM_ID } from '@angular/core';
import { NgxMasonryComponent } from './ngx-masonry.component';
import { isPlatformBrowser } from '@angular/common';
/**
 * @record
 */
function MutationWindow() { }
if (false) {
    /** @type {?} */
    MutationWindow.prototype.MutationObserver;
    /** @type {?} */
    MutationWindow.prototype.WebKitMutationObserver;
}
var NgxMasonryDirective = /** @class */ (function () {
    function NgxMasonryDirective(_element, _parent, platformId) {
        this._element = _element;
        this._parent = _parent;
        this.platformId = platformId;
    }
    /**
     * @return {?}
     */
    NgxMasonryDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platformId)) {
            this._parent.add(this._element.nativeElement);
            this.watchForHtmlChanges();
        }
    };
    /**
     * @return {?}
     */
    NgxMasonryDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platformId)) {
            this._parent.remove(this._element.nativeElement);
        }
    };
    /** When HTML in brick changes dinamically, observe that and change layout */
    /**
     * When HTML in brick changes dinamically, observe that and change layout
     * @private
     * @return {?}
     */
    NgxMasonryDirective.prototype.watchForHtmlChanges = /**
     * When HTML in brick changes dinamically, observe that and change layout
     * @private
     * @return {?}
     */
    function () {
        MutationObserver = window.MutationObserver || window.WebKitMutationObserver;
        if (MutationObserver) {
            /**
             * Watch for any changes to subtree
             * @type {?}
             */
            var self_1 = this;
            /** @type {?} */
            var observer = new MutationObserver((/**
             * @param {?} mutations
             * @param {?} observerFromElement
             * @return {?}
             */
            function (mutations, observerFromElement) {
                self_1._parent.layout();
            }));
            // define what element should be observed by the observer
            // and what types of mutations trigger the callback
            observer.observe(this._element.nativeElement, {
                subtree: true,
                childList: true
            });
        }
    };
    NgxMasonryDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxMasonryItem], ngxMasonryItem'
                },] }
    ];
    /** @nocollapse */
    NgxMasonryDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: NgxMasonryComponent, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return NgxMasonryComponent; })),] }] },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return NgxMasonryDirective;
}());
export { NgxMasonryDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxMasonryDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    NgxMasonryDirective.prototype._parent;
    /**
     * @type {?}
     * @private
     */
    NgxMasonryDirective.prototype.platformId;
}
